cpp/src/arrow/util/async_generator.h

```diff
@@ -1149,108 +1148,197 @@ class BackgroundGenerator {
     } else {
       auto next = Future<T>::MakeFinished(std::move(state_->queue.front()));
       state_->queue.pop();
-      if (!state_->running &&
-          static_cast<int>(state_->queue.size()) <= state_->q_restart) {
-        state_->RestartTask(state_, std::move(guard));
+      if (state_->NeedsRestart()) {
+        return state_->RestartTask(state_, std::move(guard), std::move(next));
       }
       return next;
     }
-    if (!state_->running) {
-      // This branch should only be needed to start the background thread on the first
-      // call
-      state_->RestartTask(state_, std::move(guard));
+    // This should only trigger the very first time this method is called
+    if (state_->NeedsRestart()) {
+      return state_->RestartTask(state_, std::move(guard), std::move(waiting_future));
     }
     return waiting_future;
   }
 
  protected:
+  enum class BackgroundThreadState : char { Reading = 0, Quitting = 1, Idle = 2 };
```

Comment:
This enum seems to be unused in favor of the booleans still?

Reply:
Ah, whoops.  Missed removing that when I was abandoning the enum approach.  I've removed it.